### ISSUE TITLE
Event loop blockage caused by newrelic harvester

### ISSUE THREAD

**Issue by DullReferenceException** [newrelic/node-newrelic#300]
I'm using [`blocked-at`](https://www.npmjs.com/package/blocked-at) to investigate event loop blockage that's been occurring in our app. Recently, I encountered an egregious sample where it was blocked for 16.8 seconds!! Call stack:

```
    at new TimersList (timers.js:201:31),
    at insert (timers.js:183:27),
    at exports.active (timers.js:152:3),
    at setTimeout (timers.js:415:3),
    at wrapper (/src/node_modules/newrelic/lib/shim/shim.js:946:24),
    at Agent._scheduleHarvester (/src/node_modules/newrelic/lib/agent.js:613:26),
    at _finish (/src/node_modules/newrelic/lib/agent.js:402:13),
    at afterHarvest (/src/node_modules/newrelic/lib/agent.js:395:7),
    at afterAllHarvestSteps (/src/node_modules/newrelic/lib/harvest.js:655:7),
    at /src/node_modules/newrelic/node_modules/async/dist/async.js:1140:9,
    at /src/node_modules/newrelic/node_modules/async/dist/async.js:473:16,
    at iterateeCallback (/src/node_modules/newrelic/node_modules/async/dist/async.js:992:24),
    at /src/node_modules/newrelic/node_modules/async/dist/async.js:969:16,
    at /src/node_modules/newrelic/node_modules/async/dist/async.js:1137:13,
    at Immediate.afterFinalize [as _onImmediate] (/src/node_modules/newrelic/lib/harvest.js:621:11),
    at runCallback (timers.js:705:18),
    at tryOnImmediate (timers.js:676:5),
    at processImmediate (timers.js:658:5),
    at process.topLevelDomainCallback (domain.js:120:23)
```

Note that this is the call stack for the beginning of the synchronous code that runs at the beginning of an async event. Based on this call stack, it's probably the `agent.harvest` code that runs after the `_scheduleHarvester` timer that's doing this.

Just a theory here, since we've had similar problems ourselves, but could this relate to the serializing of errors? We have some hefty errors flowing through our system, and we're causing event loop blockage ourselves when we try to stringify. If this is the case, what's the best way to avoid the issue? Would `error_collector: attributes: { enabled: false }` do it?

**Comment by astormnewrelic** [comment]
@DullReferenceException Thank you!  That's a pretty neat NPM package -- we should probably find a way to put something like directly into our agent :)  That 16.8 seconds does seem egregious. 

Looking over that stack trace nothing obvious sticks out.  

Re:

> but could this relate to the serializing of errors?

I'm not familiar with reading `blocked-at`'s output so I'm not 100% sure what parts of the eventLoop or micro-task-queue process it will and will not show, but I don't see anything related to errors being serialized in that stack trace?  Could you expand on your theory a little to help me understand? (or, also, just give your configuration a try and see if that sorts things out? ;) )

Also, FWIW, we're in the middle of an active refactor that's going to change _a bunch_ of things about our harvesting code  -- so we're more than happy to keep chatting about this but it may end up being a moot point.

**Comment by DullReferenceException** [comment]
> I'm not familiar with reading blocked-at's output so I'm not 100% sure what parts of the eventLoop or micro-task-queue process it will and will not show

The call stack is confusing, but it's pointing to the beginning of the synchronous code that took that amount of time. Basically, the timer callback that was set by `_scheduleHarvester` fired, so this happened during whatever synchronously happened starting here:

https://github.com/newrelic/node-newrelic/blob/8bc36123871d7ea2ab30bf4093be25ba926a187f/lib/agent.js#L616

> FWIW, we're in the middle of an active refactor that's going to change a bunch of things about our harvesting code -- so we're more than happy to keep chatting about this but it may end up being a moot point.

OK, hopefully it does end up being moot. Hopefully this puts the event loop blockage bug in your ear while you're doing the refactor so you can notice CPU-bound operations that might cause this sort of thing.

**Comment by astormnewrelic** [comment]
@DullReferenceException Hello!  We just released version 5.13 of the agent, which includes the aforementioned harvest refactoring.  We'd be curious if this clears up the event loop blocking you sometimes saw with the previous version of the agent.

**Comment by astormnewrelic** [comment]
@DullReferenceException -- we're doing a bit of GitHub cleanup today and we're going to close this issue out -- we'd still love to hear about your experience with the latest agent and whether you're still seeing event loop blocking from long harvests.
